File: lib-mvpbase-appcompat/src/main/java/com/baozi/mvp/base/BaseFragment.java
package com.baozi.mvp.base;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Looper;
import android.os.MessageQueue;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import com.baozi.mvp.presenter.BasePresenter;
import com.baozi.mvp.view.BaseFragmentView;


/**
 * @author jlanglang  2016/8/5 9:42
 */
public abstract class BaseFragment<T extends BasePresenter> extends Fragment
        implements BaseFragmentView {
    protected T mPresenter;
    protected Context mContext;//activity的上下文对象
    protected Bundle mBundle;
    private SparseArray<View> mViews;
    private View mContentView;
    private boolean isInit;


    /**
     * 绑定activity
     *
     * @param context
     */
    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mContext = context;
        //应该只创建一次Presenter.
        if (mPresenter == null || !isInit) {
            mPresenter = initPresenter();
        }
        mPresenter.onAttach(this);
    }

    /**
     * 运行在onAttach之后
     * 可以接受别人传递过来的参数,实例化对象.
     *
     * @param savedInstanceState
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //获取bundle,并保存起来
        if (savedInstanceState != null) {
            mBundle = savedInstanceState.getBundle("bundle");
        } else {
            mBundle = getArguments() == null ? new Bundle() : getArguments();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        if (mBundle != null) {
            outState.putBundle("bundle", mBundle);
        }
        super.onSaveInstanceState(outState);
    }

    /**
     * 运行在onCreate之后
     * 生成view视图
     *
     * @param inflater
     * @param container
     * @param savedInstanceState
     * @return
     */
    @Nullable
    @Override
    public final View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (null == mContentView) {
            mContentView = initView(inflater, savedInstanceState);
        } else {
            //缓存的ContentView需要判断是否已有parent， 如果有parent需要从parent删除，否则会抛出异常。
            ViewGroup parent = (ViewGroup) mContentView.getParent();
            if (parent != null) {
                parent.removeView(mContentView);
            }
        }
        return mContentView;
    }

    /**
     * 运行在onCreateView之后
     * 加载数据,初始化Presenter
     *
     * @param savedInstanceState
     */
    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        //View做一些初始化操作.
        init(savedInstanceState);
        //初始化Presenter,应该只初始化一次
        if (mPresenter != null && !isInit) {
            isInit = true;
            mPresenter.onCreate();
            //可见时再加载数据刷新视图
            Looper.myQueue().addIdleHandler(new MessageQueue.IdleHandler() {
                @Override
                public boolean queueIdle() {
                    mPresenter.initData();
                    return false;
                }
            });
            onPresentersCreate();
        }
    }

    /**
     * 扩展除了默认的presenter的其他Presenter初始化
     */
    protected void onPresentersCreate() {

    }

    @Override
    public void onStart() {
        mPresenter.onStart();
        super.onStart();
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (null != mPresenter) {
            if (isVisibleToUser) {
                //相当于Fragment的onResume
                mPresenter.onResume();
            } else {
                //相当于Fragment的onPause
                mPresenter.onPause();
            }
        }
    }

    @Override
    public void onStop() {
        mPresenter.onStop();
        super.onStop();
    }

    @Override
    public void onDetach() {
        mPresenter.onDetach();
        super.onDetach();
    }

    @Override
    public void onDestroy() {
        mPresenter.onDestroy();
        super.onDestroy();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        mPresenter.onActivityResult(requestCode, resultCode, data);
    }

    public View getContentView() {
        return mContentView;
    }


    /**
     * 类似Activity的OnBackgress
     * fragment进行回退
     */
    @Override
    public void onBack() {
        getFragmentManager().popBackStackImmediate();
    }


    /**
     * 创建Fragment视图
     *
     * @return Fragment视图
     */
    protected abstract View initView(LayoutInflater inflater, @Nullable Bundle savedInstanceState);

    /**
     * 运行在initView之后
     * 此时已经setContentView
     * 可以做一些初始化操作
     */
    public void init(Bundle savedInstanceState) {

    }

    /**
     * 跳转fragment
     *
     * @param tofragment
     */
    @Override
    public void startFragment(Fragment tofragment) {
        startFragment(tofragment, null);
    }
    /**
     * @param tofragment 跳转的fragment
     * @param tag        fragment的标签
     */
    @Override
    public void startFragment(Fragment tofragment, String tag) {
        FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
        fragmentTransaction.hide(this).add(android.R.id.content, tofragment, tag);
        fragmentTransaction.addToBackStack(tag);
        fragmentTransaction.commitAllowingStateLoss();
    }

    /**
     * 跳转Activity
     */
    public void startActivity(Class zclass) {
        Intent intent = new Intent(mContext, zclass);
        startActivity(intent);
    }

    /**
     * 跳转Activity
     */
    public void startActivity(Class zclass, Bundle bundle, int flag) {
        Intent intent = new Intent(mContext, zclass);
        intent.putExtras(bundle);
        intent.addFlags(flag);
        startActivity(intent);
    }

    /**
     * 跳转Activity
     */
    public void startActivity(Class zclass, int flag) {
        Intent intent = new Intent(mContext, zclass);
        intent.addFlags(flag);
        startActivity(intent);
    }

    @Override
    public Context getContext() {
        return mContext;
    }

    @Override
    public Bundle getBundle() {
        return mBundle;
    }

    @Override
    public BaseFragment getFragment() {
        return this;
    }

    @Override
    public Window getWindow() {
        return getActivity() == null ? ((Activity) mContext).getWindow() : getActivity().getWindow();
    }

    @Override
    public AppCompatActivity getAppcompatActivity() {
        return (AppCompatActivity) mContext;
    }

    @Override
    public ActionBar getSupportActionBar() {
        return getAppcompatActivity().getSupportActionBar();
    }

    @Override
    public void setSupportActionBar(@Nullable Toolbar toolbar) {
        getAppcompatActivity().setSupportActionBar(toolbar);
    }

    public SparseArray<View> getViews() {
        if (mViews == null) {
            mViews = new SparseArray<>();
        }
        return mViews;
    }

    /**
     * 通过viewId获取控件
     *
     * @param viewId 资源id
     * @return
     */
    public <V extends View> V findView(int viewId) {
        View view = getViews().get(viewId);
        if (view == null) {
            view = mContentView.findViewById(viewId);
            getViews().put(viewId, view);
        }
        return (V) view;
    }

    /**
     * 初始化Presenter
     *
     * @return
     */
    protected abstract T initPresenter();


}

File: lib-mvpbase-kotlin/src/main/java/com/linfeng/mvp/base/BaseFragment.kt
package com.linfeng.mvp.base

import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.os.Bundle
import android.os.Looper
import android.support.annotation.LayoutRes
import android.support.v4.app.Fragment
import android.support.v7.app.ActionBar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import com.linfeng.mvp.annotation.JMvpContract
import com.linfeng.mvp.presenter.BasePresenter
import com.linfeng.mvp.property.PresenterProperty
import com.linfeng.mvp.view.UIView
import java.lang.reflect.ParameterizedType

/**
 * @author jlanglang  2016/1/5 9:42
 */
abstract class BaseFragment<T : BasePresenter<*>> : Fragment(), UIView {


    override lateinit var mContext: Context
    protected open var mPresenter: T by PresenterProperty(this)
    protected open lateinit var mBundle: Bundle
    private var mContentView: View? = null
    private var isInit: Boolean = false
    var isFinish: Boolean = false
        get() = isDetached

    /**
     * 绑定activity

     * @param context
     */
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
        //应该只创建一次Presenter.
        if (!isInit) {
            mPresenter = initPresenter()
        }
    }

    /**
     * 运行在onAttach之后
     * 可以接受别人传递过来的参数,实例化对象.
     * @param savedInstanceState
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //获取bundle,并保存起来
        mBundle = if (savedInstanceState != null) {
            savedInstanceState.getBundle("bundle")
        } else {
            if (arguments == null) Bundle() else arguments!!
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBundle("bundle", mBundle)
        super.onSaveInstanceState(outState)
    }

    /**
     * 运行在onCreate之后
     * 生成view视图

     * @param inflater
     * *
     * @param container
     * *
     * @param savedInstanceState
     * *
     * @return
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (null == mContentView) {
            mContentView = initView(inflater, savedInstanceState)
        } else {
            //缓存的ContentView需要判断是否已有parent， 如果有parent需要从parent删除，否则会抛出异常。
            val parent = mContentView?.parent as? ViewGroup
            parent?.removeView(mContentView)
        }
        return mContentView
    }

    /**
     * 运行在onCreateView之后
     * 加载数据,初始化Presenter

     * @param savedInstanceState
     */
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        //View做一些初始化操作.
        init(savedInstanceState)
        //初始化Presenter,应该只初始化一次
        if (!isInit) {
            isInit = true
            mPresenter.onCreate()
            //可见时再加载数据刷新视图
            Looper.myQueue().addIdleHandler {
                mPresenter.onRefreshData()
                false
            }
        }
    }


    override fun onStart() {
        mPresenter.onStart()
        super.onStart()
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser) {
            //相当于Fragment的onResume
            mPresenter.onResume()
        } else {
            //相当于Fragment的onPause
            mPresenter.onPause()
        }
    }

    override fun onStop() {
        mPresenter.onStop()
        super.onStop()
    }

    override fun onDetach() {
        mPresenter.onDetach()
        super.onDetach()
    }

    override fun onDestroy() {
        mPresenter.onDestroy()
        super.onDestroy()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        mPresenter.onActivityResult(requestCode, resultCode, data)
    }

    override fun getContentView(): View {
        return mContentView!!
    }


    /**
     * 类似Activity的OnBackgress
     * fragment进行回退
     */
    override fun onBack() {
        fragmentManager?.popBackStackImmediate()
    }


    /**
     * 初始化ContentView
     * 建议不要包含toolbar

     * @param inflater
     * *
     * @param savedInstanceState
     * *
     * @return
     */
    protected open fun initView(inflater: LayoutInflater, savedInstanceState: Bundle?): View? {
        val layout = initView(savedInstanceState)
        return inflater.inflate(layout, null)
    }

    /**
     * 建议不要包含toolbar

     * @param savedInstanceState
     * *
     * @return 布局layout
     */
    @LayoutRes
    protected open fun initView(savedInstanceState: Bundle?): Int {
        val annotation = this.javaClass.getAnnotation(JMvpContract::class.java)
        if (annotation != null) {
            return annotation.layout
        }
        return 0
    }

    /**
     * 运行在initView之后
     * 此时已经setContentView
     * 可以做一些初始化操作
     */
    open fun init(savedInstanceState: Bundle?) {

    }

    override fun setSupportActionBar(toolbar: Toolbar?) {
        appcompatActivity().setSupportActionBar(toolbar)
    }

    override fun resources(): Resources = mContext.resources

    override fun appcompatActivity(): AppCompatActivity = mContext as AppCompatActivity

    override fun window(): Window = appcompatActivity().window

    override fun supportActionBar(): ActionBar? = appcompatActivity().supportActionBar

    override fun onNewThrowable(throwable: Throwable) {
    }

    override fun startFragment(fragment: Fragment) {
        startFragment(fragment, null)

    }

    override fun startFragment(fragment: Fragment, tag: String?, enterAnim: Int, exitAnim: Int, popEnter: Int, popExit: Int, isAddBack: Boolean) {
        val fragmentTransaction = fragmentManager?.beginTransaction() ?: return
        fragmentTransaction.setCustomAnimations(enterAnim, exitAnim, popEnter, popExit)
        fragmentTransaction.hide(this).add(android.R.id.content, fragment, tag)
        if (isAddBack) {
            fragmentTransaction.addToBackStack(tag)
        }
        fragmentTransaction.commitAllowingStateLoss()
    }

    override fun finishActivity() {
        appcompatActivity().finish()
    }

    /**
     * 跳转Activity
     */
    open fun startActivity(zclass: Class<*>) {
        val intent = Intent(mContext, zclass)
        startActivity(intent)
    }

    /**
     * 跳转Activity
     */
    open fun startActivity(zclass: Class<*>, bundle: Bundle, flag: Int) {
        val intent = Intent(mContext, zclass)
        intent.putExtras(bundle)
        intent.addFlags(flag)
        startActivity(intent)
    }

    /**
     * 跳转Activity
     */
    open fun startActivity(zclass: Class<*>, flag: Int) {
        val intent = Intent(mContext, zclass)
        intent.addFlags(flag)
        startActivity(intent)
    }

    /**
     * 初始化Presenter

     * @return
     */

    protected open fun initPresenter(): T {
        // 通过反射机制获取子类传递过来的实体类的类型信息
        val type = this.javaClass.genericSuperclass as ParameterizedType
        return (type.actualTypeArguments[0] as Class<T>).newInstance()
    }
}
